contracts/0.4.24/Lido.sol

```diff
@@ -79,33 +97,145 @@ contract Lido is ILido, IsContract, StETH, AragonApp {
+    * @dev A way to pause stake without pushing PAUSE for the whole proto.
+    * The main goal is to prevent huge APR losses for existing stakers due to high demands
+    * on post-Merge entry queue.
+    * Emits `StakingPaused` event.
+    */
+    function pauseStaking() external {
+        _auth(STAKING_PAUSE_ROLE);
+
+        STAKE_LIMIT_POSITION.setStorageUint256(0);
+        emit StakingPaused();
+    }
+
+    /**
+    * @notice Resume staking if `pauseStaking` was called previously (allow new submits transactions)
+    * or if new rate-limit params are required.
+    *
+    * Staking could be rate-limited by imposing a limit on the stake amount
+    * at each moment in time.
+    *
+    * ▲ Stake limit
+    * │.....  .....   ........ ...            ....     ... Stake limit = max
+    * │      .       .        .   .   .      .    . . .
+    * │     .       .              . .  . . .      . .
+    * │            .                .  . . .
+    * │──────────────────────────────────────────────────> Time
+    * │     ^      ^          ^   ^^^  ^ ^ ^     ^^^ ^     Stake events
+    *
+    * To disable rate-limit pass zero arg values.
+    * @dev Reverts if:
+    * - `_maxStakeLimit` < `_stakeLimitIncreasePerBlock`
+    * - `_maxStakeLimit` / `_stakeLimitIncreasePerBlock` >= 2^32 (only if `_stakeLimitIncreasePerBlock` != 0)
+    * Emits `StakeResumed` event
+    * @param _maxStakeLimit max stake limit value
+    * @param _stakeLimitIncreasePerBlock stake limit increase per single block
+    */
+    function resumeStaking(
+        uint96 _maxStakeLimit,
+        uint96 _stakeLimitIncreasePerBlock
+    ) external {
+        _auth(STAKING_RESUME_ROLE);
+
+        require(_maxStakeLimit >= _stakeLimitIncreasePerBlock, "TOO_LARGE_INCREASE");
+        require(
+            (_stakeLimitIncreasePerBlock == 0)
+            || (_maxStakeLimit / _stakeLimitIncreasePerBlock <= uint32(-1)),
+            "TOO_SMALL_INCREASE"
+        );
+
+        (
+            uint96 maxStakeLimit,,
+            uint96 prevStakeLimit,
+        ) = STAKE_LIMIT_POSITION.getStorageUint256().decodeStakeLimitSlot();
+
+        // if staking was paused or unlimited previously,
+        // or new limit is lower than previous, then
+        // reset prev stake limit to max
+        if ((maxStakeLimit == 0) || (_maxStakeLimit < prevStakeLimit)) {
+            prevStakeLimit = _maxStakeLimit;
+        }
+
+        STAKE_LIMIT_POSITION.setStorageUint256(
+            StakeLimitUtils.encodeStakeLimitSlot(
+                _maxStakeLimit,
+                _stakeLimitIncreasePerBlock,
+                prevStakeLimit,
+                uint32(block.number)
+            )
+        );
+
+        emit StakingResumed(_maxStakeLimit, _stakeLimitIncreasePerBlock);
+    }
+
+    /**
+    * @notice Check staking state: whether it's paused or not
+    */
+    function isStakingPaused() external view returns (bool) {
+        return STAKE_LIMIT_POSITION.getStorageUint256().isStakingPaused();
```

Comment:
(not critical, can leave as is)

Reply:
Initially, I was enthusiastic about the structures. But then it turned out that `pragma experimental ABIEncoderV2` needed for the `Lido.sol` contract to freely pass and return struct args since we stick to the solidity 0.4 here.

Reply:
I like this approach

Reply:
It's a cool concept, I am on it. :+1:

Reply:
How has it influenced the gas costs of `_submit()`?